File: lib/xrc/actions.ex
defmodule Xrc.Actions do
  @moduledoc """
  High-level operations shared by the CLI and the interactive TUI. Each function
  takes a loaded `Config` and prints user-facing progress. These compose the
  lower-level modules (Api, ExercismCli, Git, VSCode, Local).
  """

  alias Xrc.{Api, Config, ExercismCli, Git, Local, Status, VSCode}

  @doc "Download (if needed) and open an exercise in VS Code. force? re-downloads stubs."
  def start(%Config{} = config, track, exercise, force? \\ false) do
    with :ok <- sync_before_work(config) do
      already = Local.downloaded?(config, track, exercise)

      cond do
        already and not force? ->
          info("Already downloaded; continuing.")
          open_in_vscode(config, track, exercise)

        true ->
          info("Downloading #{track}/#{exercise} ...")

          case ExercismCli.download(config, track, exercise) do
            {:ok, dir} ->
              ok("Downloaded to #{dir}")
              open_in_vscode(config, track, exercise)

            {:error, reason} ->
              error(reason)
          end
      end
    end
  end

  @doc "Open an already-downloaded exercise; download first if missing."
  def open(%Config{} = config, track, exercise) do
    if Local.downloaded?(config, track, exercise) do
      open_in_vscode(config, track, exercise)
    else
      info("Not downloaded yet — fetching first.")
      start(config, track, exercise, false)
    end
  end

  defp open_in_vscode(config, track, exercise) do
    case VSCode.open(config, track, exercise) do
      :ok ->
        ok("Opened #{track}/#{exercise} in VS Code.")

      {:error, :code_not_found} ->
        dir = Local.exercise_dir(config, track, exercise)
        info("`code` not on PATH. Open manually: #{dir}")
    end
  end

  @doc "Run the exercise's tests, streaming output."
  def test(%Config{} = config, track, exercise) do
    if Local.downloaded?(config, track, exercise) do
      info("Running tests for #{track}/#{exercise} ...")

      case ExercismCli.test(config, track, exercise) do
        :ok -> ok("Tests passed.")
        {:error, code} -> error("Tests failed (exit #{code}).")
      end
    else
      error("Exercise not downloaded. Start it first.")
    end
  end

  @doc """
  Test, submit, then commit + push. `confirm` is a 1-arg fun (prompt -> bool)
  used for the resubmit and commit confirmations; defaults to auto-yes for the
  non-interactive CLI path. Pass a real prompt fun from the TUI.
  """
  def submit(%Config{} = config, track, exercise, confirm \\ fn _ -> true end) do
    cond do
      not Local.downloaded?(config, track, exercise) ->
        error("Exercise not downloaded. Start it first.")

      Git.already_completed?(config, track, exercise) and
          not confirm.("This exercise was completed before. Resubmit?") ->
        info("Aborted.")

      true ->
        run_submit(config, track, exercise, confirm)
    end
  end

  defp run_submit(config, track, exercise, confirm) do
    info("Running tests before submit ...")

    case ExercismCli.test(config, track, exercise) do
      :ok ->
        ok("Tests passed. Submitting ...")

        case ExercismCli.submit(config, track, exercise) do
          {:ok, out} ->
            ok(String.trim(out))
            commit(config, track, exercise, confirm)

          {:error, reason} ->
            error(reason)
        end

      {:error, code} ->
        error("Tests failed (exit #{code}). Fix before submitting.")
    end
  end

  defp commit(config, track, exercise, confirm) do
    info("Committing to git ...")

    result =
      Git.commit_and_push(config, track, exercise, fn stat ->
        IO.puts("\nChanges to commit:\n#{stat}")
        confirm.("Commit and push these changes?")
      end)

    case result do
      :ok -> ok("Committed and pushed: #{track}: complete #{exercise}")
      :no_changes -> info("No file changes to commit (submission still sent).")
      :aborted -> info("Commit aborted; nothing pushed.")
      {:error, reason} -> error(reason)
    end
  end

  @doc "Open the exercise's web page in the browser."
  def web(%Config{} = config, track, exercise) do
    case Api.exercises(config, track) do
      {:ok, exercises} ->
        case Enum.find(exercises, &(&1.slug == exercise)) do
          nil -> error("Unknown exercise #{exercise} in #{track}.")
          ex -> open_url(ex.web_url)
        end

      {:error, reason} ->
        error("Could not fetch exercise: #{inspect(reason)}")
    end
  end

  @doc "Open an arbitrary URL in the default browser (macOS `open`, Linux `xdg-open`)."
  def open_url(url) do
    opener =
      cond do
        System.find_executable("open") -> "open"
        System.find_executable("xdg-open") -> "xdg-open"
        true -> nil
      end

    case opener do
      nil -> info("Open manually: #{url}")
      cmd -> System.cmd(cmd, [url]) && ok("Opened #{url}")
    end
  end

  @doc "git pull --ff-only before work, requiring a clean tree (like start-exercise.sh)."
  def sync_before_work(%Config{} = config) do
    cond do
      not Git.clean?(config) ->
        error("Working tree has uncommitted changes. Commit or stash first.")
        :error

      true ->
        case Git.pull_ff(config) do
          :ok ->
            :ok

          {:error, out} ->
            error("git pull --ff-only failed:\n#{out}")
            :error
        end
    end
  end

  # --- pretty output ---

  defp ok(msg), do: Owl.IO.puts([Owl.Data.tag("✔ ", :green), msg])
  defp info(msg), do: Owl.IO.puts([Owl.Data.tag("• ", :blue), msg])
  defp error(msg), do: Owl.IO.puts([Owl.Data.tag("✘ ", :red), msg])

  @doc false
  def status_badge(status), do: Status.badge(status)
end

File: internal/actions/actions.go
// Package actions composes the lower-level exercism operations into the
// user-facing flows shared by the CLI and the TUI, printing progress as it goes.
package actions

import (
	"fmt"
	"os/exec"
	"runtime"
	"strings"

	"github.com/AirmanBugs/exercism/xrc/internal/config"
	"github.com/AirmanBugs/exercism/xrc/internal/exercism"
	"github.com/charmbracelet/lipgloss"
)

var (
	okStyle   = lipgloss.NewStyle().Foreground(lipgloss.Color("2"))
	infoStyle = lipgloss.NewStyle().Foreground(lipgloss.Color("4"))
	errStyle  = lipgloss.NewStyle().Foreground(lipgloss.Color("1"))
)

func ok(msg string)   { fmt.Println(okStyle.Render("✔ ") + msg) }
func info(msg string) { fmt.Println(infoStyle.Render("• ") + msg) }
func fail(msg string) { fmt.Println(errStyle.Render("✘ ") + msg) }

// ConfirmFunc asks a yes/no question and returns the answer.
type ConfirmFunc func(question string) bool

// AutoYes is the default confirmer for non-interactive paths.
func AutoYes(string) bool { return true }

// Start downloads (if needed) and opens an exercise in VS Code. force re-downloads stubs.
func Start(cfg *config.Config, track, exercise string, force bool) {
	if !syncBeforeWork(cfg) {
		return
	}

	if exercism.Downloaded(cfg, track, exercise) && !force {
		info("Already downloaded; continuing.")
		openEditor(cfg, track, exercise)
		return
	}

	info(fmt.Sprintf("Downloading %s/%s ...", track, exercise))
	dir, err := exercism.Download(cfg, track, exercise)
	if err != nil {
		fail(err.Error())
		return
	}
	ok("Downloaded to " + dir)
	openEditor(cfg, track, exercise)
}

// Open continues an exercise: pull first (recover WIP synced from another
// device), open it if present, otherwise download the stub. This is what makes
// "continue" work for a server-started exercise that isn't on this machine.
func Open(cfg *config.Config, track, exercise string) {
	if exercism.Downloaded(cfg, track, exercise) {
		openEditor(cfg, track, exercise)
		return
	}
	// Not on disk: it may be a WIP committed on another device, or a fresh
	// server-started exercise. Sync, then re-check before falling back to download.
	if !syncBeforeWork(cfg) {
		return
	}
	if exercism.Downloaded(cfg, track, exercise) {
		ok("Recovered work-in-progress from sync.")
		openEditor(cfg, track, exercise)
		return
	}
	info("Nothing local to continue — downloading the stub.")
	dir, err := exercism.Download(cfg, track, exercise)
	if err != nil {
		fail(err.Error())
		return
	}
	ok("Downloaded to " + dir)
	openEditor(cfg, track, exercise)
}

// Pause commits the exercise's work-in-progress and pushes it, so it can be
// resumed on another device.
func Pause(cfg *config.Config, track, exercise string) {
	if !exercism.Downloaded(cfg, track, exercise) {
		fail("Exercise not downloaded; nothing to pause.")
		return
	}
	info("Saving work-in-progress ...")
	res, err := exercism.WipCommitAndPush(cfg, track, exercise)
	if err != nil {
		fail(err.Error())
		return
	}
	switch res {
	case exercism.CommitOK:
		ok(fmt.Sprintf("Synced WIP: %s: wip %s", track, exercise))
	case exercism.CommitNoChanges:
		info("No changes to sync.")
	}
}

func openEditor(cfg *config.Config, track, exercise string) {
	err := exercism.OpenVSCode(cfg, track, exercise)
	switch {
	case err == nil:
		ok(fmt.Sprintf("Opened %s/%s in VS Code.", track, exercise))
	case err == exercism.ErrCodeNotFound:
		info("`code` not on PATH. Open manually: " + exercism.ExerciseDir(cfg, track, exercise))
	default:
		fail(err.Error())
	}

	// Show the instructions in the browser, where they render properly and can sit
	// beside the editor window (VS Code's CLI can't arrange a README pane).
	OpenURL(instructionsURL(track, exercise))
}

// instructionsURL is the exercise's instructions page on exercism.org. The URL
// is deterministic, so no API call is needed.
func instructionsURL(track, exercise string) string {
	return fmt.Sprintf("https://exercism.org/tracks/%s/exercises/%s", track, exercise)
}

// Test runs the exercise's tests, streaming output.
func Test(cfg *config.Config, track, exercise string) {
	if !exercism.Downloaded(cfg, track, exercise) {
		fail("Exercise not downloaded. Start it first.")
		return
	}
	info(fmt.Sprintf("Running tests for %s/%s ...", track, exercise))
	if err := exercism.Test(cfg, track, exercise); err != nil {
		fail("Tests failed.")
		return
	}
	ok("Tests passed.")
}

// Submit tests, submits, then commits + pushes. confirm gates resubmit/commit.
func Submit(cfg *config.Config, track, exercise string, confirm ConfirmFunc) {
	if !exercism.Downloaded(cfg, track, exercise) {
		fail("Exercise not downloaded. Start it first.")
		return
	}
	if exercism.AlreadyCompleted(cfg, track, exercise) &&
		!confirm("This exercise was completed before. Resubmit?") {
		info("Aborted.")
		return
	}

	info("Running tests before submit ...")
	if err := exercism.Test(cfg, track, exercise); err != nil {
		fail("Tests failed. Fix before submitting.")
		return
	}
	ok("Tests passed. Submitting ...")

	out, err := exercism.Submit(cfg, track, exercise)
	if err != nil {
		fail(err.Error())
		return
	}
	if out != "" {
		ok(out)
	}
	commit(cfg, track, exercise, confirm)
}

func commit(cfg *config.Config, track, exercise string, confirm ConfirmFunc) {
	info("Committing to git ...")
	res, err := exercism.CommitAndPush(cfg, track, exercise, func(stat string) bool {
		fmt.Println("\nChanges to commit:\n" + stat)
		return confirm("Commit and push these changes?")
	})
	if err != nil {
		fail(err.Error())
		return
	}
	switch res {
	case exercism.CommitOK:
		ok(fmt.Sprintf("Committed and pushed: %s: complete %s", track, exercise))
	case exercism.CommitNoChanges:
		info("No file changes to commit (submission still sent).")
	case exercism.CommitAborted:
		info("Commit aborted; nothing pushed.")
	}
}

// Web opens the exercise's web page in the browser.
func Web(cfg *config.Config, track, exercise string) {
	client := exercism.NewClient(cfg)
	exs, err := client.Exercises(track)
	if err != nil {
		fail("Could not fetch exercise: " + err.Error())
		return
	}
	for _, e := range exs {
		if e.Slug == exercise {
			OpenURL(e.WebURL)
			return
		}
	}
	fail(fmt.Sprintf("Unknown exercise %s in %s.", exercise, track))
}

// OpenURL opens a URL in the default browser.
func OpenURL(url string) {
	var cmd string
	switch runtime.GOOS {
	case "darwin":
		cmd = "open"
	default:
		cmd = "xdg-open"
	}
	if _, err := exec.LookPath(cmd); err != nil {
		info("Open manually: " + url)
		return
	}
	_ = exec.Command(cmd, url).Run()
	ok("Opened " + url)
}

func syncBeforeWork(cfg *config.Config) bool {
	if !exercism.GitClean(cfg) {
		fail("Working tree has uncommitted changes. Commit or stash first.")
		return false
	}
	if err := exercism.GitPullFF(cfg); err != nil {
		fail("git pull --ff-only failed:\n" + strings.TrimSpace(err.Error()))
		return false
	}
	return true
}
